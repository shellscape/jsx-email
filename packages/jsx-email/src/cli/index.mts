#!/usr/bin/env node --enable-source-maps --no-warnings=ExperimentalWarning

import './main.mjs';
