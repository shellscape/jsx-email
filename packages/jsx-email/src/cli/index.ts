#!/usr/bin/env node
import 'source-map-support/register';

import importLocal from 'import-local';
import yargs from 'yargs-parser';

import pkg from '../../package.json';
import { debug } from '../debug';

import { command as build } from './commands/build';
import { command as check } from './commands/check';
import { command as create } from './commands/create';
import { command as help } from './commands/help';
import { command as preview } from './commands/preview';
import type { CommandFn } from './commands/types';

const commands: Record<string, CommandFn> = { build, check, create, help, preview };

const { log } = console;

const run = async () => {
  const argv = yargs(process.argv.slice(2), { configuration: { 'strip-dashed': true } });
  const { _: positionals, ...flags } = argv;
  const [commandName] = positionals;
  let command = commands[commandName];

  debug.cli(`Command Name: \`${commandName}\``);

  if (flags.version) {
    log(`${pkg.name} v${pkg.version}\n`);
    return;
  }

  if (!command) command = help;

  const input = (positionals.slice(1) as string[]) || [];
  const result = await command(flags, input);

  if (!result) {
    debug.cli(`Command \`${commandName}\` returned \`false\``);
    help({}, []);
  }
};

if (importLocal(__filename)) {
  debug.cli('Using local install of webpack-command');
} else {
  run();
}
