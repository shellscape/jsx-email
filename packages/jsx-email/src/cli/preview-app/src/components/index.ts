export * from './code';
export * from './heading';
export * from './icons';
export * from './logo';
export * from './sidebar';
export * from './text';
export * from './topbar';
