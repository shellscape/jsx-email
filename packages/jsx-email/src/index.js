import './helpers.js';
// components
export * from './components/background.js';
export * from './components/body.js';
export * from './components/button.js';
export * from './components/butan.js';
export * from './components/code.js';
export * from './components/color-scheme.js';
export * from './components/column.js';
export * from './components/conditional.js';
export * from './components/container.js';
export * from './components/font.js';
export * from './components/graph.js';
export * from './components/head.js';
export * from './components/heading.js';
export * from './components/hr.js';
export * from './components/html.js';
export * from './components/img.js';
export * from './components/link.js';
export * from './components/markdown.js';
export * from './components/preview.js';
export * from './components/row.js';
export * from './components/raw.js';
export * from './components/section.js';
export * from './components/tailwind/tailwind.js';
export * from './components/text.js';
// renderer
export * from './renderer/compat/context.js';
export * from './renderer/compat/hooks.js';
export * from './renderer/jsx-to-string.js';
export * from './renderer/render.js';
export { useData } from './renderer/suspense.js';
export * from './plugins.js';
export * from './types.js';
//# sourceMappingURL=index.js.map