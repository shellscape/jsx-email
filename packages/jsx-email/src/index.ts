export * from './components/body';
export * from './components/button/button';
export * from './components/code';
export * from './components/color-scheme';
export * from './components/column';
export * from './components/conditional';
export * from './components/container';
export * from './components/font';
export * from './components/head';
export * from './components/heading';
export * from './components/hr';
export * from './components/html';
export * from './components/img';
export * from './components/link';
export * from './components/markdown';
export * from './components/preview';
export * from './components/row';
export * from './components/section';
export * from './components/tailwind/tailwind';
export * from './components/text';
export * from './render';
export * from './types';
