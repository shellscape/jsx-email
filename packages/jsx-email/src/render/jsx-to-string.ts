/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
import hash from 'hash-it';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

const promiseMap = new Map();

const wrapPromise = <TPromise extends Promise<any>>(promise: TPromise) => {
  let status = 'pending';
  let result: Awaited<TPromise>;
  const suspender = promise.then(
    (r) => {
      status = 'success';
      result = r;
    },
    (e) => {
      status = 'error';
      result = e;
    }
  );
  return {
    // eslint-disable-next-line consistent-return
    read() {
      if (status === 'pending') {
        throw suspender;
      } else if (status === 'error') {
        throw result;
      } else if (status === 'success') {
        return result;
      }
    }
  };
};

export const useData = <TData>(props: any, cb: () => Promise<TData>): TData => {
  const key = hash(props);
  let dataPromise;
  if (promiseMap.has(key)) {
    dataPromise = promiseMap.get(key);
  } else {
    dataPromise = wrapPromise(cb());
    promiseMap.set(key, dataPromise);
  }
  return dataPromise.read();
};

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error('Unsupported JSX element:', element as any);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }

    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
