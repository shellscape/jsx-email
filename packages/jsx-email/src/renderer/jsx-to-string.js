/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import { log } from '../log.js';
import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';
import { isValidElement } from './is-valid-element.js';
const renderSuspense = async (children) => {
    try {
        const result = await jsxToString(children);
        return result;
    }
    catch (error) {
        if (error instanceof Promise) {
            await error;
            return renderSuspense(children);
        }
        throw error;
    }
};
/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element) {
    if (element == null) {
        return '';
    }
    if (typeof element === 'string') {
        return escapeString(element);
    }
    else if (typeof element === 'number') {
        return String(element);
    }
    else if (typeof element === 'boolean' || element == null) {
        return '';
    }
    else if (isIterable(element)) {
        let html = '';
        for (const child of element) {
            // eslint-disable-next-line no-await-in-loop
            html += await jsxToString(child);
        }
        return html;
    }
    // Handle Promise case
    if (element instanceof Promise) {
        const resolvedElement = await element;
        return jsxToString(resolvedElement);
    }
    if (!isValidElement(element)) {
        log.error(chalk `{red Unsupported JSX element}:`, element);
        throw new Error(`Unsupported JSX element`);
    }
    const { type } = element;
    const props = element.props || EmptyObject;
    if (typeof type === 'string') {
        let html = `<${type}`;
        let innerHTML = '';
        for (const prop of Object.keys(props)) {
            const value = props[prop];
            if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
                // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
            }
            else if (prop === 'class' || prop === 'className') {
                // This condition is here because it is the most common attribute
                // and short-circuiting results in a ~5% performance boost.
                html += value ? ` class="${escapeString(value)}"` : '';
            }
            else if (prop === 'style') {
                html += ` style="${stringifyStyles(value)}"`;
            }
            else if (prop === 'dangerouslySetInnerHTML') {
                // eslint-disable-next-line no-underscore-dangle
                innerHTML = value.__html;
            }
            else {
                const name = AttributeAliases[prop] || prop;
                if (BooleanAttributes.has(name)) {
                    html += value ? ` ${name}` : '';
                }
                else if (typeof value === 'string') {
                    html += ` ${name}="${escapeString(value)}"`;
                }
                else if (typeof value === 'number') {
                    html += ` ${name}="${String(value)}"`;
                }
                else if (typeof value === 'boolean') {
                    html += ` ${name}="${value}"`;
                }
            }
        }
        if (VoidElements.has(type)) {
            html += '/>';
        }
        else {
            html += '>';
            if (innerHTML) {
                html += innerHTML;
            }
            else {
                html += await jsxToString(props.children);
            }
            html += `</${type}>`;
        }
        return html;
    }
    else if (type) {
        if (typeof type === 'function') {
            const result = type(props);
            const renderedFC = await (result instanceof Promise
                ? result.then(jsxToString)
                : jsxToString(result));
            const sym = type.$$typeof;
            if (sym && Symbol.keyFor(sym) === 'react.provider') {
                type.context.pop();
            }
            return renderedFC;
        }
        else if (typeof type === 'symbol') {
            const key = Symbol.keyFor(type);
            // is this react fragment?
            if (key === 'react.fragment') {
                return jsxToString(props.children);
            }
            else if (key === 'react.suspense') {
                const suspenseResult = await renderSuspense(props.children);
                return suspenseResult;
            }
        }
        else if (isReactForwardRef(type)) {
            return jsxToString(type.render(props, props.ref));
        }
        else if (type.$$typeof) {
            const key = Symbol.keyFor(type.$$typeof);
            if (key === 'react.provider') {
                type.context.pop();
            }
            else if (key === 'react.context') {
                return jsxToString(props.children);
            }
        }
        log.error(chalk `{red Unsupported JSX element}:`, type);
        throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
    }
    return '';
}
function isIterable(node) {
    return typeof node === 'object' && node !== null && Symbol.iterator in node;
}
function isReactForwardRef(type) {
    return (typeof type === 'object' &&
        type !== null &&
        'render' in type &&
        '$$typeof' in type &&
        typeof type.$$typeof === 'symbol' &&
        Symbol.keyFor(type.$$typeof) === 'react.forward_ref');
}
//# sourceMappingURL=jsx-to-string.js.map