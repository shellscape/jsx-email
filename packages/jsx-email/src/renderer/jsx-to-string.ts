/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import type { FC, ReactNode } from 'react';

import { log } from '../log.js';

import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    log.error(chalk`{red Unsupported JSX element}:`, element);
    throw new Error(`Unsupported JSX element`);
  }

  const { type } = element;
  const props = element.props || EmptyObject;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = AttributeAliases[prop as keyof typeof AttributeAliases] || prop;

        if (BooleanAttributes.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VoidElements.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      const renderedFC = await jsxToString((type as FC)(props));
      const sym = (type as { $$typeof?: symbol }).$$typeof;

      if (sym && Symbol.keyFor(sym) === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      }

      return renderedFC;
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    } else if ((type as { $$typeof?: symbol }).$$typeof) {
      const key = Symbol.keyFor((type as { $$typeof: symbol }).$$typeof);
      if (key === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      } else if (key === 'react.context') {
        return jsxToString(props.children);
      }
    }

    log.error(chalk`{red Unsupported JSX element}:`, type);
    throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
