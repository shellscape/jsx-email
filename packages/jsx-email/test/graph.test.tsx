// @ts-ignore
import React from 'react';

import { jsxToString } from '../src/renderer/jsx-to-string.js';
import { Graph } from '../src';

describe('<Hr> component', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('passes styles and other props correctly', async () => {
    const html = await jsxToString(
      <Graph
        width={300}
        height={300}
        title="Basic graph demo"
        config={{
          data: {
            datasets: [
              {
                data: [50, 60, 70, 180],
                label: 'Users'
              }
            ],
            labels: ['Q1', 'Q2', 'Q3', 'Q4']
          },
          type: 'bar'
        }}
      />
    );
    expect(html).toContain('width="300"');
    expect(html).toContain('height="300"');
    expect(html).toContain('src="https://quickchart.io/chart?');
  });

  it('renders correctly', async () => {
    const actualOutput = await jsxToString(
      <Graph
        title="Basic graph demo"
        config={{
          data: {
            datasets: [
              {
                data: [50, 60, 70, 180],
                label: 'Users'
              }
            ],
            labels: ['Q1', 'Q2', 'Q3', 'Q4']
          },
          type: 'bar'
        }}
      />
    );
    expect(actualOutput).toMatchSnapshot();
  });
});
