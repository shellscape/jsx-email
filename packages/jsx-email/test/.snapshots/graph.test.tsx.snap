// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Hr> component > renders correctly 1`] = `"<img src=\\"https://quickchart.io/chart?c=%7B%22data%22%3A%7B%22datasets%22%3A%5B%7B%22data%22%3A%5B50%2C60%2C70%2C180%5D%2C%22label%22%3A%22Users%22%7D%5D%2C%22labels%22%3A%5B%22Q1%22%2C%22Q2%22%2C%22Q3%22%2C%22Q4%22%5D%7D%2C%22type%22%3A%22bar%22%7D\\" alt=\\"Basic graph demo\\"/>"`;
